File: swift/token.swift
enum Token {
    case plus
    case minus
    case number(Double)
    case identifier(String)

    func lexeme() -> String {
        switch self {
            case .plus: return "+"
            case .minus: return "-"
            case .number(let value): return String(value)
            case .identifier(let name): return name
        }
    }
}

let expression: [Token] = [.number(4.0), .plus, .identifier("x")]
let text = expression.map{$0.lexeme()}.joined(separator: " ")
assert(text == "4.0 + x")

File: rust/src/bin/token.rs
enum Token {
    Plus,
    Minus,
    Number(f64),
    Identifier(String),
}

impl Token {
    fn lexeme(&self) -> String {
        match self {
            Token::Plus => String::from("+"),
            Token::Minus => String::from("-"),
            Token::Number(value) => value.to_string(),
            Token::Identifier(name) => name.clone(),
        }
    }
}

fn main() {
    let expression: Vec<Token> = vec![
        Token::Number(3.0),
        Token::Plus,
        Token::Identifier("x".to_string()),
    ];
    let text = expression.iter()
        .map(|t| t.lexeme())
        .collect::<Vec<String>>()
        .join(" ");
    assert_eq!(text, "3 + x");
}
